hail/src/main/scala/is/hail/cxx/Emit.scala

```diff
@@ -19,6 +21,160 @@ abstract class ArrayEmitter(val setup: Code, val m: Code, val setupLen: Code, va
   def emit(f: (Code, Code) => Code): Code
 }
 
+class Orderings {
+  val orderings = new ArrayBuilder[Code]
+  val typeOrdering = mutable.Map.empty[(PType, PType), String]
+
+  private def makeOrdering(tub: TranslationUnitBuilder, lp: PType, rp: PType): String = {
+    lp.virtualType match {
+      case TFloat32(_) => "FloatOrd"
+      case TFloat64(_) => "DoubleOrd"
+      case TInt32(_) => "IntOrd"
+      case TInt64(_) => "LongOrd"
+      case TBoolean(_) => "BoolOrd"
+
+      case TString(_) | TBinary(_) => "BinaryOrd"
+
+      case t: TContainer =>
+        val lContainerP = lp.asInstanceOf[PContainer]
+        val rContainerP = rp.asInstanceOf[PContainer]
+
+        val lElemP = lContainerP.elementType
+        val rElemP = rContainerP.elementType
+
+        val elemOrd = ordering(tub, lContainerP.elementType, rContainerP.elementType)
+        val eElemOrd = s"ExtOrd<$elemOrd>"
+
+        s"ArrayOrd<${ lContainerP.cxxImpl },${ rContainerP.cxxImpl },ExtOrd<$elemOrd>>"
+
+      case t: TBaseStruct =>
+        val ord = tub.genSym("Ord")
+
+        val lBaseStructP = lp.asInstanceOf[PBaseStruct]
+        val rBaseStructP = rp.asInstanceOf[PBaseStruct]
+
+        def comparisonTemplate(
+          op: String,
+          rType: String,
+          fieldTest: (TranslationUnitBuilder, String, Code, Code, Code, Code) => Code,
+          finalValue: Code): Code = {
+          s"""
+             |static $rType $op(const char *l, const char *r) {
+             |  ${
+            t.fields.zipWithIndex.map { case (f, i) =>
+              val fOrd = ordering(tub, lBaseStructP.fields(i).typ, rBaseStructP.fields(i).typ)
+
+              val lm = tub.variable("lm", "bool", lBaseStructP.cxxIsFieldMissing("l", i))
+              val lv = tub.variable("lv", typeToCXXType(lBaseStructP.fields(i).typ))
+
+              val rm = tub.variable("rm", "bool", rBaseStructP.cxxIsFieldMissing("r", i))
+              val rv = tub.variable("lv", typeToCXXType(rBaseStructP.fields(i).typ))
+
+              s"""
+                 |${lm.define}
+                 |${lv.define}
+                 |if (!$lm)
+                 |  $lv = ${lBaseStructP.cxxLoadField("l", i)};
+                 |${rm.define}
+                 |${rv.define}
+                 |if (!$rm)
+                 |  $rv = ${rBaseStructP.cxxLoadField("r", i)};
+                 |${ fieldTest(tub, s"ExtOrd<$fOrd>", lm.toString, lv.toString, rm.toString, rv.toString) }
+               """.stripMargin
+            }.mkString
+          }
+             |  return $finalValue;
+             |}
+           """.stripMargin
+        }
+
+        orderings +=
+          s"""
+             |class $ord {
```

Comment:
should this be written using functions/classes instead? Do you prefer to just generate strings inline like this?

Reply:
I think it might look neater if comparisonTemplate returned a function and then `ord` is just a class with the six ordering functions and the `using T = const char *;` statement

Reply:
Totally agree, fixed.